zap/src/output/client.rs

```diff
@@ -1,260 +1,363 @@
-use std::fmt::Display;
-
 use crate::{
-	irgen::{gen_des, gen_ser},
-	parser::{EvCall, EvSource, EvType, File},
+	irgen::{gen_des, gen_ser, Stmt},
+	parser::{EvCall, EvDecl, EvSource, EvType, File, TyDecl},
 	util::casing,
 };
 
-struct ClientFile<'a>(&'a File);
+use super::Output;
+
+struct ClientOutput<'a> {
+	file: &'a File,
+	buff: String,
+	tabs: u32,
+}
+
+impl<'a> Output for ClientOutput<'a> {
+	fn push(&mut self, s: &str) {
+		self.buff.push_str(s);
+	}
+
+	fn indent(&mut self) {
+		self.tabs += 1;
+	}
+
+	fn dedent(&mut self) {
+		self.tabs -= 1;
+	}
+
+	fn push_indent(&mut self) {
+		for _ in 0..self.tabs {
+			self.push("\t");
+		}
+	}
+}
```

Comment:
These methods should be inherited from the trait.

Reply:
this isn't possible in rust as traits cannot declare fields

Reply:
Seems like it is https://doc.rust-lang.org/book/ch10-02-traits.html ?